### PR TITLE
Add ruff pre-commit-hook

### PR DESCRIPTION
**Pull request by waltsims**
This PR adds a pre-commit hook for ruff linting and runs it in a larger update commit.

### CODE REVIEW

**Comment by waltsims** [comment]
- [ ] add developer documentation on how to install pre-commit

File: examples/at_array_as_source/at_array_as_source.py [hunk 1]
```diff
@@ -79,17 +79,17 @@ def main():
 
     # combine source signals into one array
     source_signal = np.zeros((3, max(len(sig1), len(sig2))))
-    source_signal[0, :len(sig1)] = sig1
-    source_signal[1, :len(sig2)] = sig2
-    source_signal[2, :len(sig3)] = sig3
+    source_signal[0, : len(sig1)] = sig1
+    source_signal[1, : len(sig2)] = sig2
+    source_signal[2, : len(sig3)] = sig3
```
Comment (faridyagubbayli) [inline examples/at_array_as_source/at_array_as_source.py:84]: Not sure if we want to have a space here?

File: examples/at_linear_array_transducer/at_linear_array_transducer.py [hunk 2]
```diff
@@ -49,11 +49,10 @@ def main():
     else:
         ids = np.arange(1, element_num + 1) - (element_num + 1) / 2
 
-    time_delays = -(np.sqrt((ids * element_pitch) ** 2 + source_focus ** 2) - source_focus) / c0
+    time_delays = -(np.sqrt((ids * element_pitch) ** 2 + source_focus**2) - source_focus) / c0
```
Comment (faridyagubbayli) [inline examples/at_linear_array_transducer/at_linear_array_transducer.py:52]: Why did it remove the whitespace around second `**` operator while whitespace around the first is left?

File: examples/us_bmode_linear_transducer/example_utils.py [hunk 3]
```diff
@@ -2,12 +2,10 @@
 
 
 def download_from_gdrive(file_id, output_path):
-
     try:
         import gdown
     except ModuleNotFoundError:
-        raise AssertionError('This example requires `gdown` to be installed. '
-                             'Please install using `pip install gdown`')
+        raise AssertionError("This example requires `gdown` to be installed. " "Please install using `pip install gdown`")
```
Comment (faridyagubbayli) [inline examples/us_bmode_linear_transducer/example_utils.py:8]:
```suggestion
        raise AssertionError("This example requires `gdown` to be installed. Please install using `pip install gdown`")
```

File: kwave/executor.py [hunk 4]
```diff
@@ -21,31 +19,32 @@ def _make_binary_executable(self):
...
-                  f'-i {input_filename} ' \
-                  f'-o {output_filename} ' \
-                  f'{options}'
+        command = (
+            f"{self.execution_options.system_string} "
+            f"{self.execution_options.binary_path} "
+            f"-i {input_filename} "
+            f"-o {output_filename} "
+            f"{options}"
+        )
```
Comment (faridyagubbayli) [inline kwave/executor.py:28]: Enclosing multi-line string with paranthesis, interesting ...

File: kwave/kWaveSimulation_helper/scale_source_terms_func.py [hunk 5]
```diff
@@ -122,14 +123,14 @@ def scale_pressure_source_dirichlet(source_p, c0, N, p_source_pos_index):
     if c0.size == 1:
         # compute the scale parameter based on the homogeneous
         # sound speed
-        source_p = source_p / (N * (c0 ** 2))
+        source_p = source_p / (N * (c0**2))
```
Comment (faridyagubbayli) [inline kwave/kWaveSimulation_helper/scale_source_terms_func.py:126]: It sometimes adds and sometimes removes whitespaces around `**`

File: tests/test_us_bmode_phased_array.py [hunk 6]
```diff
@@ -160,14 +158,10 @@ def test_us_bmode_phased_array():
...
     y_pos = grid_spacing_meters.y * (grid_size_points.y / 2)  # [m]
-    ball_location = Vector([
-        round(x_pos / grid_spacing_meters.x + rounding_eps),
-        round(y_pos / grid_spacing_meters.x + rounding_eps),
-        grid_size_points.z / 2
-    ])
-    scattering_region1 = make_ball(
-        grid_size_points, ball_location, round(radius / grid_spacing_meters.x + rounding_eps)
+    ball_location = Vector(
+        [round(x_pos / grid_spacing_meters.x + rounding_eps), round(y_pos / grid_spacing_meters.x + rounding_eps), grid_size_points.z / 2]
```
Comment (faridyagubbayli) [inline tests/test_us_bmode_phased_array.py:162]: I'd prefer this to stay as it was before

File: tests/test_sd_sensor_directivity_2D.py [hunk 7]
```diff
@@ -77,15 +77,11 @@ def test_sd_sensor_directivity_2D():
...
+    input_filename = "example_def_tran_input.h5"
     pathname = gettempdir()
     input_file_full_path = os.path.join(pathname, input_filename)
     simulation_options = SimulationOptions(
-        pml_alpha=np.array([2, 0]),
-        save_to_disk=True,
-        input_filename=input_filename,
-        data_path=pathname,
-        save_to_disk_exit=True
+        pml_alpha=np.array([2, 0]), save_to_disk=True, input_filename=input_filename, data_path=pathname, save_to_disk_exit=True
```
Comment (faridyagubbayli) [inline tests/test_sd_sensor_directivity_2D.py:84]: I think one-param per line reads better

**Review by faridyagubbayli — APPROVED** [review]
Some comments mainly about my preferences. We can discuss and if necessary adress them later. They are not blockers. I think this hook will help a lot to standardize code style in the repo. Thanks a lot for working on this!

**Comment by faridyagubbayli** [comment]
Merging this PR will be difficult with ongoing development due to large number of changed files. I propose to freeze development during weekly meeting, rebasing this PR on `main`, running the pre-commit hook and merging the PR.

**Pull request merged**